src/lib/transaction.rs

```diff
@@ -0,0 +1,586 @@
+                                "Could not deserialize nullifier",
+                            ))?,
+                    );
+
+                    Ok(n)
+                })
+            })
+            .collect::<Result<Vec<usize>, io::Error>>()?;
+
+        n += (&buf[n..]).read(&mut one_byte)?;
+        if u8::from_le_bytes(one_byte) != 0 {
+            let mut crossover = Crossover::default();
+            n += crossover.write(&buf[n..])?;
+            self.crossover = Some(crossover);
+        }
+
+        n += (&buf[n..]).read(&mut one_u64)?;
+        let notes_size = u64::from_le_bytes(one_u64) as usize;
+        self.notes = vec![Note::default(); notes_size];
+        self.notes
+            .iter_mut()
+            .map(|note| {
+                (&buf[n..]).read(&mut one_note).and_then(|num| {
+                    n += num;
+                    note.write(&one_note).and_then(|_| Ok(num))
+                })
+            })
+            .collect::<Result<Vec<usize>, io::Error>>()?;
+
+        n += self.fee.write(&buf[n..])?;
+
+        let mut proof_data = vec![0u8; PROOF_SIZE];
+        n += (&buf[n..]).read(&mut proof_data)?;
+        let proof = Proof::from_bytes(&proof_data).map_err(|_| {
+            io::Error::new(
+                io::ErrorKind::InvalidData,
+                "Could not deserialize proof",
+            )
+        })?;
+        self.spending_proof = proof;
+
+        n += (&buf[n..]).read(&mut one_u64)?;
+        let call_data_size = u64::from_le_bytes(one_u64) as usize;
+        let mut call_data = vec![0u8; call_data_size];
+        n += (&buf[n..]).read(&mut call_data)?;
+        self.call_data = call_data;
+
+        Ok(n)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::services::rusk_proto;
+    use dusk_pki::PublicSpendKey;
+    use dusk_plonk::bls12_381::Scalar as BlsScalar;
+    use dusk_plonk::jubjub::{
+        AffinePoint as JubJubAffine, ExtendedPoint as JubJubExtended,
+        Fr as JubJubScalar, GENERATOR_EXTENDED,
+    };
+    use phoenix_core::Note;
+    use rand::Rng;
+    use std::convert::TryInto;
+    use std::io::{Read, Write};
+
+    fn random_note() -> Note {
+        let t: u8 = rand::thread_rng().gen_range(0, 2);
+
+        let s = JubJubScalar::random(&mut rand::thread_rng());
+        let a = GENERATOR_EXTENDED * s;
+
+        let s = JubJubScalar::random(&mut rand::thread_rng());
+        let b = GENERATOR_EXTENDED * s;
+
+        // We need to cast extended points to affine and back,
+        // to ensure integrity of data.
+        let a = JubJubExtended::from(JubJubAffine::from(a));
+        let b = JubJubExtended::from(JubJubAffine::from(b));
+
+        let pk = PublicSpendKey::new(a, b);
+
+        let value: u64 = rand::thread_rng().gen();
+
+        Note::new(t.try_into().unwrap(), &pk, value)
+    }
+
+    fn random_fee() -> Fee {
```

Comment:
As discussing, let's limit the randomness of tests to those that really needs it, otherwise is hard to have tests results that are always reproducible.